File: apple/AestheticAlgorithm/Algorithms/Sorting/HeapSort.swift
import Foundation

struct HeapSort: AlgorithmExecutable {
    let id = "heap-sort"
    let name = "Heap Sort"
    let category = "sorting"
    let timeComplexity = (best: "O(n log n)", average: "O(n log n)", worst: "O(n log n)")
    let spaceComplexity = "O(1)"

    let pseudocode: [(line: Int, text: String)] = [
        (0, "procedure heapSort(A)"),
        (1, "  n ← length(A)"),
        (2, "  // Build max heap"),
        (3, "  for i ← n/2 - 1 downto 0 do"),
        (4, "    heapify(A, n, i)"),
        (5, "  // Extract elements from heap"),
        (6, "  for i ← n - 1 downto 1 do"),
        (7, "    swap(A[0], A[i])"),
        (8, "    heapify(A, i, 0)"),
        (9, ""),
        (10, "procedure heapify(A, n, i)"),
        (11, "  largest ← i"),
        (12, "  left ← 2*i + 1"),
        (13, "  right ← 2*i + 2"),
        (14, "  if left < n and A[left] > A[largest]"),
        (15, "    largest ← left"),
        (16, "  if right < n and A[right] > A[largest]"),
        (17, "    largest ← right"),
        (18, "  if largest ≠ i then"),
        (19, "    swap(A[i], A[largest])"),
        (20, "    heapify(A, n, largest)")
    ]

    func generateSteps(from input: [Int]) -> [AlgorithmStep] {
        var array = input
        var steps: [AlgorithmStep] = []
        var sortedIndices: [Int] = []
        let n = array.count

        if n == 0 { return steps }

        // Initial state
        steps.append(AlgorithmStep(
            type: .highlight,
            array: array,
            highlightedIndices: [],
            secondaryIndices: [],
            sortedIndices: [],
            pseudocodeLine: 0,
            description: "Starting heap sort on array of \(n) elements"
        ))

        // Build max heap phase
        steps.append(AlgorithmStep(
            type: .highlight,
            array: array,
            highlightedIndices: [],
            secondaryIndices: [],
            sortedIndices: [],
            pseudocodeLine: 2,
            description: "Phase 1: Building max heap"
        ))

        for i in stride(from: n / 2 - 1, through: 0, by: -1) {
            heapify(&array, heapSize: n, rootIndex: i, steps: &steps, sortedIndices: &sortedIndices)
        }

        // Max heap built
        steps.append(AlgorithmStep(
            type: .highlight,
            array: array,
            highlightedIndices: Array(0..<n),
            secondaryIndices: [],
            sortedIndices: sortedIndices,
            pseudocodeLine: 4,
            description: "Max heap built: \(array)"
        ))

        // Extraction phase
        steps.append(AlgorithmStep(
            type: .highlight,
            array: array,
            highlightedIndices: [],
            secondaryIndices: [],
            sortedIndices: sortedIndices,
            pseudocodeLine: 5,
            description: "Phase 2: Extracting elements from heap"
        ))

        for i in stride(from: n - 1, through: 1, by: -1) {
            // Swap root (max) with last unsorted element
            steps.append(AlgorithmStep(
                type: .compare,
                array: array,
                highlightedIndices: [0, i],
                secondaryIndices: [],
                sortedIndices: sortedIndices,
                pseudocodeLine: 7,
                description: "Moving max element \(array[0]) to position \(i)"
            ))

            array.swapAt(0, i)

            steps.append(AlgorithmStep(
                type: .swap,
                array: array,
                highlightedIndices: [0, i],
                secondaryIndices: [],
                sortedIndices: sortedIndices,
                pseudocodeLine: 7,
                description: "Swapped \(array[0]) and \(array[i])"
            ))

            // Mark as sorted
            sortedIndices.append(i)
            steps.append(AlgorithmStep(
                type: .sorted,
                array: array,
                highlightedIndices: [i],
                secondaryIndices: [],
                sortedIndices: sortedIndices,
                pseudocodeLine: 7,
                description: "\(array[i]) is now in its final sorted position"
            ))

            // Restore heap property for reduced heap
            heapify(&array, heapSize: i, rootIndex: 0, steps: &steps, sortedIndices: &sortedIndices)
        }

        // First element is sorted by default
        sortedIndices.append(0)
        steps.append(AlgorithmStep(
            type: .sorted,
            array: array,
            highlightedIndices: Array(0..<n),
            secondaryIndices: [],
            sortedIndices: Array(0..<n),
            pseudocodeLine: 0,
            description: "Heap sort complete"
        ))

        return steps
    }

    private func heapify(_ array: inout [Int], heapSize: Int, rootIndex: Int, steps: inout [AlgorithmStep], sortedIndices: inout [Int]) {
        var largest = rootIndex
        let left = 2 * rootIndex + 1
        let right = 2 * rootIndex + 2

        // Compare with left child
        if left < heapSize {
            steps.append(AlgorithmStep(
                type: .compare,
                array: array,
                highlightedIndices: [largest, left],
                secondaryIndices: [],
                sortedIndices: sortedIndices,
                pseudocodeLine: 14,
                description: "Comparing \(array[largest]) with left child \(array[left])"
            ))

            if array[left] > array[largest] {
                largest = left
            }
        }

        // Compare with right child
        if right < heapSize {
            steps.append(AlgorithmStep(
                type: .compare,
                array: array,
                highlightedIndices: [largest, right],
                secondaryIndices: [],
                sortedIndices: sortedIndices,
                pseudocodeLine: 16,
                description: "Comparing \(array[largest]) with right child \(array[right])"
            ))

            if array[right] > array[largest] {
                largest = right
            }
        }

        // Swap if needed and recurse
        if largest != rootIndex {
            steps.append(AlgorithmStep(
                type: .swap,
                array: array,
                highlightedIndices: [rootIndex, largest],
                secondaryIndices: [],
                sortedIndices: sortedIndices,
                pseudocodeLine: 19,
                description: "Swapping \(array[rootIndex]) and \(array[largest]) to maintain heap property"
            ))

            array.swapAt(rootIndex, largest)

            // Recursively heapify the affected subtree
            heapify(&array, heapSize: heapSize, rootIndex: largest, steps: &steps, sortedIndices: &sortedIndices)
        }
    }
}

File: web/src/algorithms/sorting/heap-sort.ts
import type { AlgorithmImplementation, AlgorithmStep } from '@/lib/types/algorithm';

export const heapSort: AlgorithmImplementation = {
  id: 'heap-sort',
  name: 'Heap Sort',
  category: 'sorting',
  timeComplexity: { best: 'O(n log n)', average: 'O(n log n)', worst: 'O(n log n)' },
  spaceComplexity: 'O(1)',
  pseudocode: [
    { line: 0, text: 'procedure heapSort(A)' },
    { line: 1, text: '  buildMaxHeap(A)' },
    { line: 2, text: '  for i \u2190 n-1 downto 1 do' },
    { line: 3, text: '    swap A[0] and A[i]' },
    { line: 4, text: '    heapify(A, 0, i)' },
    { line: 5, text: 'procedure heapify(A, i, size)' },
    { line: 6, text: '  largest \u2190 i' },
    { line: 7, text: '  left \u2190 2*i + 1' },
    { line: 8, text: '  right \u2190 2*i + 2' },
    { line: 9, text: '  if left < size and A[left] > A[largest]' },
    { line: 10, text: '    largest \u2190 left' },
    { line: 11, text: '  if right < size and A[right] > A[largest]' },
    { line: 12, text: '    largest \u2190 right' },
    { line: 13, text: '  if largest \u2260 i then' },
    { line: 14, text: '    swap A[i] and A[largest]' },
    { line: 15, text: '    heapify(A, largest, size)' },
  ],

  generateSteps(input: number[]): AlgorithmStep[] {
    const array = [...input];
    const steps: AlgorithmStep[] = [];
    const n = array.length;
    const sortedIndices: number[] = [];

    function heapify(size: number, i: number): void {
      let largest = i;
      const left = 2 * i + 1;
      const right = 2 * i + 2;

      // Show heapify start
      steps.push({
        type: 'highlight',
        array: [...array],
        highlightedIndices: [i],
        secondaryIndices: [
          ...(left < size ? [left] : []),
          ...(right < size ? [right] : []),
        ],
        sortedIndices: [...sortedIndices],
        pseudocodeLine: 6,
        description: `Heapify at index ${i} (value ${array[i]})`,
      });

      if (left < size) {
        steps.push({
          type: 'compare',
          array: [...array],
          highlightedIndices: [left, largest],
          secondaryIndices: [],
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 9,
          description: `Compare left child ${array[left]} with ${array[largest]}`,
        });

        if (array[left] > array[largest]) {
          largest = left;
        }
      }

      if (right < size) {
        steps.push({
          type: 'compare',
          array: [...array],
          highlightedIndices: [right, largest],
          secondaryIndices: [],
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 11,
          description: `Compare right child ${array[right]} with ${array[largest]}`,
        });

        if (array[right] > array[largest]) {
          largest = right;
        }
      }

      if (largest !== i) {
        [array[i], array[largest]] = [array[largest], array[i]];
        steps.push({
          type: 'swap',
          array: [...array],
          highlightedIndices: [i, largest],
          secondaryIndices: [],
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 14,
          description: `Swap ${array[i]} and ${array[largest]}`,
        });

        heapify(size, largest);
      }
    }

    // Build max heap
    steps.push({
      type: 'pass-complete',
      array: [...array],
      highlightedIndices: [],
      secondaryIndices: [],
      sortedIndices: [],
      pseudocodeLine: 1,
      description: 'Building max heap',
    });

    for (let i = Math.floor(n / 2) - 1; i >= 0; i--) {
      heapify(n, i);
    }

    steps.push({
      type: 'pass-complete',
      array: [...array],
      highlightedIndices: [],
      secondaryIndices: [],
      sortedIndices: [...sortedIndices],
      pseudocodeLine: 1,
      description: 'Max heap built',
    });

    // Extract elements from heap
    for (let i = n - 1; i > 0; i--) {
      // Swap root (max) with last unsorted element
      [array[0], array[i]] = [array[i], array[0]];
      steps.push({
        type: 'swap',
        array: [...array],
        highlightedIndices: [0, i],
        secondaryIndices: [],
        sortedIndices: [...sortedIndices],
        pseudocodeLine: 3,
        description: `Swap max element ${array[i]} with ${array[0]}`,
      });

      // Mark as sorted
      sortedIndices.push(i);
      steps.push({
        type: 'sorted',
        array: [...array],
        highlightedIndices: [i],
        secondaryIndices: [],
        sortedIndices: [...sortedIndices],
        pseudocodeLine: 3,
        description: `${array[i]} is now in its final position`,
      });

      // Restore heap property
      heapify(i, 0);
    }

    // Mark the last remaining element as sorted
    sortedIndices.push(0);
    steps.push({
      type: 'sorted',
      array: [...array],
      highlightedIndices: [0],
      secondaryIndices: [],
      sortedIndices: [...sortedIndices],
      pseudocodeLine: 0,
      description: `${array[0]} is now in its final position. Array is fully sorted.`,
    });

    return steps;
  },
};
